### PR TITLE
fixed polymorphism issues

### PR DESCRIPTION
**Pull request by williexu**
-adds polymorphic discriminator as long as the base model is polymorphic.

### CODE REVIEW

**Comment by azuresdkci1x** [comment]
# 🤖 AutoRest automatic publish job 🤖
~~~ Haskell
> will publish once PR gets merged
~~~

File: test/vanilla/package.json [hunk 1]
```diff
@@ -20,12 +20,12 @@
   "dependencies": {
     "@types/mocha": "^2.2.39",
     "@types/should": "^8.1.30",
-    "@types/node": "^6.0.66",
+    "@types/node": "^8.0.53",
```
Comment (olydis) [inline test/vanilla/package.json]: very bueno (y)

**Comment by olydis** [comment]
> win32-x64 restart

**Comment by olydis** [comment]
@williexu ignore my comments here, just testing my coverage report generation tool - NodeJS seems to be missing some tests, so this is PR is a good test case for me 😉

**Comment by olydis** [comment]
> win32-x64 restart

**Comment by olydis** [comment]
# 🤖 AutoRest automatic feature coverage report 🤖
*feature set version 2.3.9*

## General: 98% 

The following 10 features are not covered by your tests:
❌ [`allowedValueEnum`](https://github.com/Azure/autorest.testserver/search?q=allowedValueEnum)
❌ [`expectedEnum`](https://github.com/Azure/autorest.testserver/search?q=expectedEnum)
❌ [`getComplexPolymorphismComplicated`](https://github.com/Azure/autorest.testserver/search?q=getComplexPolymorphismComplicated)
❌ [`getEnumReferenced`](https://github.com/Azure/autorest.testserver/search?q=getEnumReferenced)
❌ [`getEnumReferencedConstant`](https://github.com/Azure/autorest.testserver/search?q=getEnumReferencedConstant)
❌ [`putComplexPolymorphismComplicated`](https://github.com/Azure/autorest.testserver/search?q=putComplexPolymorphismComplicated)
❌ [`putEnumReferenced`](https://github.com/Azure/autorest.testserver/search?q=putEnumReferenced)
❌ [`putEnumReferencedConstant`](https://github.com/Azure/autorest.testserver/search?q=putEnumReferencedConstant)
❌ [`roundTripEnum`](https://github.com/Azure/autorest.testserver/search?q=roundTripEnum)
❌ [`unexpectedEnum`](https://github.com/Azure/autorest.testserver/search?q=unexpectedEnum)


## Azure: 0% 

The following 1 features are not covered by your tests:
❌ [``](https://github.com/Azure/autorest.testserver/search?q=)

**Comment by olydis** [comment]
thanks @cormacpayne!

**Comment by olydis** [comment]
@williexu BTW, I'd probably disable that test that checks for coverage (as long as those enum tests aren't covered) - it seems the NodeJS Azure tests aren't even run because of the early failure, so by ignoring the missing coverage within the tests... you get more coverage. 😄 the "AutoRest automatic feature coverage report" will then also contain your coverage numbers for Azure, so it's not like you/Amar will ever miss that stuff is still missing 😉

**Comment by williexu** [comment]
That does sounds more ideal. Thanks :)

**Comment by azuresdkci1x** [comment]
# 🤖 AutoRest automatic publish job 🤖
~~~ Haskell
> will publish once PR gets merged
~~~

**Comment by azuresdkci1x** [comment]
# 🤖 AutoRest automatic publish job 🤖
~~~ Haskell
> will publish once PR gets merged
~~~

**Comment by olydis** [comment]
# 🤖 AutoRest automatic feature coverage report 🤖
*feature set version 2.3.9*

## General: 0% 

❌ no tests were run for this category


## Azure: 0% 

❌ no tests were run for this category

**Comment by olydis** [comment]
# 🤖 AutoRest automatic feature coverage report 🤖
*feature set version 2.3.9*

## General: 0% 

❌ no tests were run for this category


## Azure: 0% 

❌ no tests were run for this category

**Comment by olydis** [comment]
@amarzavery @williexu SHOOT, actually please add the coverage test back in - the test server knows that you're finished with your tests by waiting for you to ask for the coverage report... so please just make that test case not throw if the coverage isn't perfect. Other than that, the tests only seem to fail because of a diff, so a final regenerate run should make this PR ready.

**Comment by williexu** [comment]
will do

**Comment by olydis** [comment]
# 🤖 AutoRest automatic feature coverage report 🤖
*feature set version 2.3.9*

## General: 0% 

❌ no tests were run for this category


## Azure: 0% 

❌ no tests were run for this category

**Comment by olydis** [comment]
# 🤖 AutoRest automatic feature coverage report 🤖
*feature set version 2.3.9*

## General: 0% 

❌ no tests were run for this category


## Azure: 0% 

❌ no tests were run for this category

**Comment by olydis** [comment]
# 🤖 AutoRest automatic feature coverage report 🤖
*feature set version 2.3.9*

## General: 98% 

The following 10 features are not covered by your tests:
❌ [`allowedValueEnum`](https://github.com/Azure/autorest.testserver/search?q=allowedValueEnum)
❌ [`expectedEnum`](https://github.com/Azure/autorest.testserver/search?q=expectedEnum)
❌ [`getComplexPolymorphismComplicated`](https://github.com/Azure/autorest.testserver/search?q=getComplexPolymorphismComplicated)
❌ [`getEnumReferenced`](https://github.com/Azure/autorest.testserver/search?q=getEnumReferenced)
❌ [`getEnumReferencedConstant`](https://github.com/Azure/autorest.testserver/search?q=getEnumReferencedConstant)
❌ [`putComplexPolymorphismComplicated`](https://github.com/Azure/autorest.testserver/search?q=putComplexPolymorphismComplicated)
❌ [`putEnumReferenced`](https://github.com/Azure/autorest.testserver/search?q=putEnumReferenced)
❌ [`putEnumReferencedConstant`](https://github.com/Azure/autorest.testserver/search?q=putEnumReferencedConstant)
❌ [`roundTripEnum`](https://github.com/Azure/autorest.testserver/search?q=roundTripEnum)
❌ [`unexpectedEnum`](https://github.com/Azure/autorest.testserver/search?q=unexpectedEnum)


## Azure: 99% 

The following 1 features are not covered by your tests:
❌ [`LRONonRetryPut201Creating400InvalidJson`](https://github.com/Azure/autorest.testserver/search?q=LRONonRetryPut201Creating400InvalidJson)

**Comment by olydis** [comment]
# 🤖 AutoRest automatic feature coverage report 🤖
*feature set version 2.3.9*

## General: 98% 

The following 10 features are not covered by your tests:
❌ [`allowedValueEnum`](https://github.com/Azure/autorest.testserver/search?q=allowedValueEnum)
❌ [`expectedEnum`](https://github.com/Azure/autorest.testserver/search?q=expectedEnum)
❌ [`getComplexPolymorphismComplicated`](https://github.com/Azure/autorest.testserver/search?q=getComplexPolymorphismComplicated)
❌ [`getEnumReferenced`](https://github.com/Azure/autorest.testserver/search?q=getEnumReferenced)
❌ [`getEnumReferencedConstant`](https://github.com/Azure/autorest.testserver/search?q=getEnumReferencedConstant)
❌ [`putComplexPolymorphismComplicated`](https://github.com/Azure/autorest.testserver/search?q=putComplexPolymorphismComplicated)
❌ [`putEnumReferenced`](https://github.com/Azure/autorest.testserver/search?q=putEnumReferenced)
❌ [`putEnumReferencedConstant`](https://github.com/Azure/autorest.testserver/search?q=putEnumReferencedConstant)
❌ [`roundTripEnum`](https://github.com/Azure/autorest.testserver/search?q=roundTripEnum)
❌ [`unexpectedEnum`](https://github.com/Azure/autorest.testserver/search?q=unexpectedEnum)


## Azure: 99% 

The following 1 features are not covered by your tests:
❌ [`LRONonRetryPut201Creating400InvalidJson`](https://github.com/Azure/autorest.testserver/search?q=LRONonRetryPut201Creating400InvalidJson)

**Comment by olydis** [comment]
Smooth, you guys have the best out-of-the-box coverage among all generators! C# now is [here](https://autorestci1.blob.core.windows.net/autorest-ci-coverage/autorest.csharp_2.2.43.md) but had over 50 cases uncovered when I first activated the report... awesome work. 🥇

**Pull request merged**